### PR TITLE
New article/natserract

### PR DESCRIPTION
**Pull request by natserract**
New article: elixir pattern match, assignment & immutability

### CODE REVIEW

File: content/elixir-assignment.md [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+twitter:
+telegram: dotnatserract
+date: 2022-06-05
+cover: https://files.speakerdeck.com/presentations/0ee235074d534893ac19cf5218a58204/slide_3.jpg
+categories:
+  - elixir
+  - functional programming
+---
+
+[Elixir](https://github.com/elixir-lang/elixir) adalah bahasa pemrogaman fungsional, dinamis, dan metaprogrammable dibangun untuk membuat aplikasi yang scalable dan maintanable. Elixir dibangun dari bahasa pemrogaman [Erlang](https://www.erlang.org/), secara bersamaan bahasa ini melakukan pendekatan [actor based](https://www.oreilly.com/library/view/functional-programming-a/9781680502756/f_0061.xhtml) untuk masalah **konkurensi** dengan syntax yang lebih sederhana, clean, namun kuat.
```
Comment (elianiva) [inline content/elixir-assignment.md:15]: lebih prefer kalo istilah yg kaya gini ga perlu ditranslate

```suggestion
[Elixir](https://github.com/elixir-lang/elixir) adalah bahasa pemrograman fungsional, dinamis, dan _metaprogrammable_ dibangun untuk membuat aplikasi yang scalable dan maintanable. Elixir dibangun dari bahasa pemrogaman [Erlang](https://www.erlang.org/), secara bersamaan bahasa ini melakukan pendekatan [actor based](https://www.oreilly.com/library/view/functional-programming-a/9781680502756/f_0061.xhtml) untuk masalah **concurrency** dengan syntax yang lebih sederhana, clean, namun kuat.
```

File: content/elixir-assignment.md [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+
+[José Valim](https://github.com/josevalim) creator dari elixir membangun bahasa ini dengan tujuan untuk meningkatkan ekstensibilitas dan produktivitas yang lebih tinggi dengan Erlang Virtual Machine yaitu [BEAM (Bogdan/Björn’s Erlang Abstract Machine)](<https://en.wikipedia.org/wiki/BEAM_(Erlang_virtual_machine)>). VM ini akan meng-compile source code Elixir dan Erlang ke file bytecode (`.beam`) yang nantinya berjalan di BEAM. Ini berarti kita bisa menjalankan erlang code di elixir tanpa ada masalah runtime sama sekali.
+
+> There’s nothing you can do in Erlang that can’t be done in Elixir!
+
+## Pattern Matching & Assignment
+
+Jika Anda mulai belajar Elixir, tentunya Anda akan bertemu dengan pernyataan yang sedikit membingungkan di Elixir, disana disebut bahwa Elixir tidak memiliki operator assignment `=`. Maka dari itu disini kita coba kupas secara mendetail bagaiamana pernyataan aneh tersebut menjadi sesuatu yang kita bisa pahami bagaimana itu bekerja.
```
Comment (elianiva) [inline content/elixir-assignment.md:23]:
```suggestion
Jika Anda mulai belajar Elixir, tentunya Anda akan bertemu dengan pernyataan yang sedikit membingungkan di Elixir, disana disebut bahwa Elixir tidak memiliki operator assignment `=`. Maka dari itu disini kita coba kupas secara mendetail bagaimana pernyataan aneh tersebut menjadi sesuatu yang kita bisa pahami bagaimana itu bekerja.
```

File: content/elixir-assignment.md [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+
+In a nutshell, pattern matching adalah pola untuk membandingkan suatu data dengan struktur pola tertentu. Cara ini digunakan di banyak bahasa pemrogaman. Dalam kutipan artikel [Wikipedia](https://en.wikipedia.org/wiki/SNOBOL), pencocokan pola pertama kali digunakan pada [SNOBOL(1962)](http://www.snobol4.org/docs/burks/tutorial/ch4.htm) yang digunakan untuk operasi string. Di masa modern kini, fitur ini banyak digunakan seperti di bahasa pemrogaman seperti OCaml, ML, Rust, Haskell, etc.
```
Comment (elianiva) [inline content/elixir-assignment.md:25]:
```suggestion
In a nutshell, pattern matching adalah pola untuk membandingkan suatu data dengan struktur pola tertentu. Cara ini digunakan di banyak bahasa pemrogaman. Dalam kutipan artikel [Wikipedia](https://en.wikipedia.org/wiki/SNOBOL), pattern matching pertama kali digunakan pada [SNOBOL(1962)](http://www.snobol4.org/docs/burks/tutorial/ch4.htm) yang digunakan untuk operasi string. Di masa modern kini, fitur ini banyak digunakan seperti di bahasa pemrograman seperti OCaml, ML, Rust, Haskell, etc.
```

File: content/elixir-assignment.md [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+
+Dalam elixir, pattern match digunakan salah satunya saat pendeklarasian variabel (bindings). **Elixir merupakan bahasa pemrogaman dinamis**, jadi Anda tidak perlu mendeklarasikan variabel atau tipenya secara _eksplisit_:
+
+```bash
```
Comment (elianiva) [inline content/elixir-assignment.md:29]:
```suggestion
```rust
```

File: content/elixir-assignment.md [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+iex(3)> a = 1
+1
+
+iex(4)> b = 2
+2
+```
+
+Itu adalah code yang simple, di bahasa pemrogaman seperti javascript itu terlihat seperti operasi [assignment](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Assignment) (`=`) nilai pada sebuah variabel. Namun ketika di Elixir, itu adalah hal yang berbeda. Di elixir, operator `=` bukanlah operator assignment melainkan _match operator_ atau bisa juga disebut _binding_.
```
Comment (elianiva) [inline content/elixir-assignment.md:55]:
```suggestion
Itu adalah contoh kode yang sederhana, di bahasa pemrograman seperti Javascript itu terlihat seperti operasi [assignment](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Assignment) (`=`) nilai pada sebuah variabel. Namun ketika di Elixir, itu adalah hal yang berbeda. Di Elixir, operator `=` bukanlah operator assignment melainkan _match operator_ atau bisa juga disebut _binding_.
```

File: content/elixir-assignment.md [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+
+Dalam elixir, apapun adalah ekspresi, dan setiap ekspresi memiliki hasil. Artinya, dalam case diatas hasilnya adalah apapun yang ada disebelah kanan operator `=` yaitu `1`, `2`. Jadi **ketika mendeklarasikan variabel dengan nilai, maka variabel tersebut sudah menjadi milik nilai itu**. Lalu antara nilai dengan variabel **di evaluasi/ di-check jika itu cocok/sama** maka hasil terakhirnya selesai dievaluasi.
```
Comment (elianiva) [inline content/elixir-assignment.md:57]:
```suggestion
Dalam Elixir, apapun adalah ekspresi, dan setiap ekspresi memiliki hasil. Artinya, dalam case diatas hasilnya adalah apapun yang ada disebelah kanan operator `=` yaitu `1`, `2`. Jadi **ketika mendeklarasikan variabel dengan nilai, maka variabel tersebut sudah menjadi milik nilai itu**. Lalu antara nilai dengan variabel **di-evaluasi / di-check jika itu cocok/sama** maka hasil terakhirnya selesai dievaluasi.
```

File: content/elixir-assignment.md [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+# ^ 5 tidak sama dengan `num`
+
+iex(2)> 10 = num
+# ^ 10 sama dengan `num`
+
+10
+# ^ Hasil dari ekspresi terakhir
+```
+
+Anda lihat? pada bagian operasi pertama `num = 10`, proses yang terjadi yakni:
```
Comment (elianiva) [inline content/elixir-assignment.md:81]:
```suggestion
Anda lihat? Pada bagian operasi pertama `num = 10`, proses yang terjadi yakni:
```

File: content/elixir-assignment.md [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+
+- Nilai 10 di cocokan dengan `num` apakah nilainya ada? jika ada maka 10 dicocokan dengan nilai tersebut. Namun jika nilai tidak ada, maka artinya `num` memiliki nilai 10. Proses pencocokan tersebut, sekaligus melakukan binding `num = 10` dengan `10 = num`.
+- Saat coba dicocokan nilai yang lain, `5 = num` dan itu sudah exists. Maka artinya 5 tidak sama dengan 10 dan tentunya `match error`.
```
Comment (elianiva) [inline content/elixir-assignment.md:84]:
```suggestion
- Saat coba dicocokkan dengan nilai yang lain, `5 = num` dan itu sudah exists. Maka artinya 5 tidak sama dengan 10 dan tentunya `match error`.
```

File: content/elixir-assignment.md [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+
+Itu sebenernya seperti kondisi biasa, jika _num == 10_ maka cetak!
```
Comment (elianiva) [inline content/elixir-assignment.md:86]:
```suggestion
Itu sebenarnya seperti kondisi biasa, jika _num == 10_ maka cetak!
```

File: content/elixir-assignment.md [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+
+Nama variabel juga dapat diakhiri dengan karakter tanda tanya (?) atau tanda seru (!):
+
+```bash
+valid_name?
+juga_ok!
+```
+
+- (`?`): Fungsi yang mengembalikan nilai boolean.
+- (`!`): Fungsi atau makro di mana kasus kegagalan memunculkan pengecualian.
```
Comment (elianiva) [inline content/elixir-assignment.md:128]:
```suggestion
- (`!`): Fungsi atau makro di mana kasus kegagalan memunculkan exception.
```

File: content/elixir-assignment.md [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+
+Lihat [konvensi penamaan](https://hexdocs.pm/elixir/main/naming-conventions.html) lainnya.
+
+### Bindings, Allocations
+
+Bind secara sederhana, **mengaitkan/memberi nama dengan suatu nilai atau fungsi**. Di elixir, binding dilakukan dengan operator `=`, disebelah kanan adalah ekspresi, disebelah kiri adalah namanya. Bind dilakukan untuk** mereferensikan suatu istilah (variabel) ke dalam memori**. Jadi ketika membuat variabel, itu secara tidak langsung menjadikan nama variabel tersebut menjadi referensi dari memori.
```
Comment (elianiva) [inline content/elixir-assignment.md:134]:
```suggestion
Bind secara sederhana, **mengaitkan/memberi nama dengan suatu nilai atau fungsi**. Di Elixir, binding dilakukan dengan operator `=`, disebelah kanan adalah ekspresi, disebelah kiri adalah namanya. Bind dilakukan untuk **mereferensikan suatu istilah (variabel) ke dalam memori**. Jadi ketika membuat variabel, itu secara tidak langsung menjadikan nama variabel tersebut menjadi referensi dari memori.
```

File: content/elixir-assignment.md [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+iex(2)> num # <- Verify
+10
+
+iex(3)> num = 20 # <- Bind ulang variabel num
+20
+iex(4)> num # <- Verify
+20
+```
+
+Secara default semua data di Elixir bersifat [immutable](https://en.wikipedia.org/wiki/Immutable_object). Artinya, yang dilakukan pada operasi diatas bukanlah merubah nilai `num` menjadi `20`, melainkan rebinding `num` dengan `20`. Rebinding di elixir, **tidak mengubah nilai aslinya**. Tetapi, itu **membuat memori baru dan menetapkan nama ke lokasi memori yang baru**. Jadi ketika rebind dilakukan, istilah (variabel) tersebut akan mengambil referensi ke lokasi memori yang berbeda. Ini berkaitan dengan [manajemen memori](https://www.honeybadger.io/blog/elixir-memory-structure/) (allocations) di Elixir.
```
Comment (elianiva) [inline content/elixir-assignment.md:150]:
```suggestion
Secara default semua data di Elixir bersifat [immutable](https://en.wikipedia.org/wiki/Immutable_object). Artinya, yang dilakukan pada operasi diatas bukanlah merubah nilai `num` menjadi `20`, melainkan rebinding `num` dengan `20`. Rebinding di Elixir **tidak mengubah nilai aslinya**. Tetapi **membuat memori baru dan menetapkan nama ke lokasi memori yang baru**. Jadi ketika rebind dilakukan, istilah (variabel) tersebut akan mengambil referensi ke lokasi memori yang berbeda. Ini berkaitan dengan [manajemen memori](https://www.honeybadger.io/blog/elixir-memory-structure/) (allocations) di Elixir.
```

File: content/elixir-assignment.md [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+
+Elixir disebut bahasa fungsional, karena evaluasi suatu program menggunakan konsep evaluasi fungsi dalam matematika. Secara bersamaan fp selalu populer dengan sifat _pure_ dan _immutable_. Immutable merupakan konsep inti Pemrograman Fungsional.
+
+Immutable berarti, **data yang sudah dibuat tidak dapat diubah**. Secara sederhana, itu berarti **semua variabel Anda adalah konstan**. Ini juga berkaitan dengan fungsi murni (_pure function_), ketika diberikan input yang sama, akan selalu menghasilkan output yang sama dan tidak memiliki efek samping.
+
+### Bekerja dengan Immutable Data
+
+Data di Elixir secara default tidak dapat diubah. Setiap fungsi akan mengembalikan **salinan baru yang dimodifikasi dari data input**. Untuk mengubahnya, Anda harus **mengambil salinan baru ke variabel lain atau re-bind kembali ke nama/variabel yang sama**.
+
+Nantinya, hasil evaluasi akan disimpan di lokasi memori lain. Jadi jika ingin me-modifikasi input tidak mengubah inputan aslinya, tetapi melalui beberapa penyalinan data. Coba lihat contoh ini:
```
Comment (elianiva) [inline content/elixir-assignment.md:166]:
```suggestion
Nantinya hasil evaluasi akan disimpan di lokasi memori lain. Jadi ketika kita me-modifikasi input maka hal itu tidak akan mengubah inputan aslinya, tetapi akan melalui beberapa penyalinan data. Coba lihat contoh ini:
```

File: content/elixir-assignment.md [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+
+iex(2)> list2 = [4 | list]
+[4, 3, 2, 1]
+
+iex(3)> list = list2
+[4, 3, 2, 1]
+#  ^ reference baru
+```
+
+Lists di elixir disimpan sebagai [linked list](https://www.educative.io/edpresso/what-is-a-linked-list) dimana setiap data (nodes) saling berhubungan dengan node berikutnya. Node pertama disebut kepala (`head`) dan terakhir disebut ekor (`tail`):
```
Comment (elianiva) [inline content/elixir-assignment.md:180]:
```suggestion
List di Elixir disimpan sebagai [linked list](https://www.educative.io/edpresso/what-is-a-linked-list) dimana setiap data (nodes) saling berhubungan dengan node berikutnya. Node pertama disebut kepala (`head`) dan terakhir disebut ekor (`tail`):
```

File: content/elixir-assignment.md [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+
+# String.capitalize(t(), :default | :ascii | :greek | :turkic) :: t()
+iex(10)> new_name = String.capitalize(name)
+"Elixir"
+
+iex(11)> name
+"elixir"
+```
+
+`String.capitalize()`, mengonversi karakter pertama dalam string menjadi huruf besar dan sisanya menjadi huruf kecil menurut mode (e.g. `:default`).
```
Comment (elianiva) [inline content/elixir-assignment.md:207]:
```suggestion
`String.capitalize()` mengkonversi karakter pertama dalam string menjadi huruf besar dan sisanya menjadi huruf kecil menurut mode (e.g. `:default`).
```

File: content/elixir-assignment.md [hunk 16]
```diff
@@ -0,0 +1,222 @@
...
+
+Ada dua manfaat penting dari immutability, yaitu **terbebas dari side effect (safe)** dan **[konsistensi data](https://en.wikipedia.org/wiki/Purely_functional_data_structure)**. Sebenernya itu saling berhubungan, terbebas dari side effect ini berarti Anda tidak perlu khawatir data Anda akan berubah, ketika satu instance dan instance lainnya mengakses data tersebut, ia harus membuat salinan baru sehingga data aslinya tidak berubah (_think if you have 1 million processes_).
+
+Immutability, sekaligus juga membuat kode Anda lebih mudah diuji, dan mudah dipahami, karena jalur nya jelas fokus bagian perbagian sesuai tugasnya. Sebenernya masih banyak manfaat dari konsep ini, tapi saya bisa berpendapat bahwa:
+
+> **Inti fokus dari pemrogaman fungsional adalah untuk konsistensi dan keamanan data. Sedangkan [OOP](https://www.techtarget.com/searchapparchitecture/definition/object-oriented-programming-OOP) sendiri lebih fokus untuk manufaktur(arsitekturing) dan desain suatu aplikasi**. Meskipun kedua paradigma ini bisa untuk dua alasan diatas, itu adalah murni pendapat saya. _So every people can make own decision right?_
```
Comment (elianiva) [inline content/elixir-assignment.md:217]:
```suggestion
> **Inti fokus dari pemrogaman fungsional adalah untuk konsistensi dan keamanan data. Sedangkan [OOP](https://www.techtarget.com/searchapparchitecture/definition/object-oriented-programming-OOP) sendiri lebih fokus untuk manufaktur (arsitektur) dan desain suatu aplikasi**. Meskipun kedua paradigma ini bisa untuk dua alasan diatas, itu adalah murni pendapat saya. _So every people can make own decision right?_
```

**Review by elianiva — CHANGES_REQUESTED** [review]
nice, cuma ada beberapa typo/stylistic changes

btw, yarn.lock hapus aja, udah ada package-lock.json

**Comment by natserract** [comment]
Thankyou for review @elianiva, updated ya please check again.

**Review by elianiva — APPROVED** [review]
sip, jangan lupa hapus yarn.lock nanti

**Review by aldy505 — APPROVED** [review]

**Pull request merged**